homeassistant/components/recorder/entity_registry.py

```diff
@@ -40,16 +39,11 @@ def entity_registry_changed_filter(
         """Handle entity_id changed filter."""
         return event_data["action"] == "update" and "old_entity_id" in event_data
 
-    @callback
-    def _setup_entity_registry_event_handler(hass: HomeAssistant) -> None:
-        """Subscribe to event registry events."""
-        hass.bus.async_listen(
-            er.EVENT_ENTITY_REGISTRY_UPDATED,
-            _async_entity_id_changed,
-            event_filter=entity_registry_changed_filter,
-        )
-
-    async_at_start(hass, _setup_entity_registry_event_handler)
+    hass.bus.async_listen(
+        er.EVENT_ENTITY_REGISTRY_UPDATED,
+        _async_entity_id_changed,
+        event_filter=entity_registry_changed_filter,
+    )
```

Comment:
It's very critical that recorder is the first to know about a renamed entity but this PR doesn't guard against someone else installing an earlier listener.

I think we need to guard against it:
```suggestion
    if er.EVENT_ENTITY_REGISTRY_UPDATED in hass.bus.async_listeners:
        # TODO: Fail recorder set up?
    
    hass.bus.async_listen(
        er.EVENT_ENTITY_REGISTRY_UPDATED,
        _async_entity_id_changed,
        event_filter=entity_registry_changed_filter,
    )
```

Reply:
Yup thats what the TODO is for.  Wasn't sure if it should make recorder setup fail at run time or log an error. Maybe patch it in the test to fail at test time.  I was worried there will be custom components out there that break the order somehow and having the recorder fail to set up might be too aggressive for this case.

Reply:
I think we should break in that case. We're setting up recorder very early, and if custom integrations patch bootstrapping  to load early they should make sure they don't load before recorder.

Reply:
https://github.com/home-assistant/core/pull/142584/commits/0cdefed4dd6059b94edf00b0398aeb12cd24f48b

Reply:
Looks like that check won't work because we have two other listeners that do get installed before which are ok
```
[(<Job listen entity_registry_updated HassJobType.Callback <bound method EntityRegistryDisabledHandler._handle_entry_updated of <homeassistant.config_entries.EntityRegistryDisabledHandler object at 0x114ef4830>>>, <function _handle_entry_updated_filter at 0x10510d8a0>), (<Job listen entity_registry_updated HassJobType.Callback <function _async_setup_entity_restore.<locals>.cleanup_restored_states at 0x114ef96c0>>, <function _async_setup_entity_restore.<locals>.cleanup_restored_states_filter at 0x114ef9620>)]
```

Reply:
Added `async_has_entity_registry_updated_listeners` instead. Probably should split that into a new PR once we are happy with this